### ISSUE TITLE
rack request environment variables are allways encoded in ascii-8bit regardless of default_external Encoding 

### ISSUE THREAD

**Issue by dre-hh** [phusion/passenger#1328]
While in a rails controller

    request.original_url.encoding
     => #<Encoding:ASCII-8BIT>

    Encoding.default_external
    => #<Encoding:UTF-8>

This dilemma would lead to 

     ActionView::Template::Error (incompatible character encodings: ASCII-8BIT and UTF-8)

when trying to render those vars in a view.

There are some closed issues regarding this problem,  e.g #971, however the error still occur.
We could reproduce this issue on several servers with proper LC_ALL, LANG etc. set to en_US.UTF-8

The request environment variables are all encoded properly in UTF-8 when running rails on a develoopment webrick

Using Phusion Passenger version 4.0.42 behind nginx/1.6.0 on Debian GNU/Linux 7.5 (wheezy)

**Comment by FooBarWidget** [comment]
@dre-hh Your premise is that the app crashes because it tries to concatenate two strings with different encodings. But what I'm saying is that that is not necessarily true. Ruby *can* concatenate a UTF-8 and a binary string, as long as the binary string only contains ASCII data.

For example, this works fine because the second string only contains ASCII data, despite being marked as a binary string.

```
>> "✓" << "ascii-as-binary".force_encoding("binary")
=> "✓ascii-as-binary"
```

But this causes a crash as you would expect (0xFF is not valid ASCII):

```
>> "✓" << "\xFF".force_encoding("binary")
Encoding::CompatibilityError: incompatible character encodings: UTF-8 and ASCII-8BIT
```

In other words, even though Passenger currently marks the Rack environment strings as binary, as long as they only contain ASCII, everything should be fine. We've already confirmed that REQUEST_URI only contains ASCII data because the URI is in escaped form.

So this leaves the question of why you experienced a crash in the first place. As things are, I don't think changing the encoding of the Rack env would solve anything; there's probably a deeper problem that we haven't explored.

I haven't been able to reproduce your particular issue. Would you be able to make a small test application that can reproduce the problem, so that I can investigate further?

**Comment by dre-hh** [comment]
I 'll try to find the time to make one. The problem is, this error was
indeed hard to reproduce. We've got reported those issues on newrelic and
on some machines within QA Team. I personally could never reproduce this
issue neither  on staging or  live deplyoments :) and with none of all
popular browsers

2015-01-02 12:20 GMT+01:00 Hongli Lai <notifications@github.com>:

**Comment by raggi** [comment]
If it's the URI, look in the load balancer logs (if you have them).

**Comment by dre-hh** [comment]
I've created a demo rails app.  It's the same scenario. Generating some social media share links.
https://github.com/dre-hh/request_env_encoding_bug

However i cannot reproduce the error on my machine. The setup is using thin (it also encodes request env in ascii ).  We also want to deploy the demo on some public server. I'll also ask then the QA team if they can reproduce the bug. 

Works on my machine! :)

**Comment by dre-hh** [comment]
http://encoding-bug.exsemt.de/

**Comment by FooBarWidget** [comment]
@dre-hh I'm unable to trigger the error with that app. What input did you use to trigger the error?

**Comment by dre-hh** [comment]
:) i can't and nobody in the QA Team can trigger it either. Thx for your time. I suggest to close it, as it's a Heisenbug

**Comment by FooBarWidget** [comment]
Too bad. :( Closing this issue until someone can find a way to reliably reproduce it.
